File: LeetCode题目集/290. Word Pattern(easy).java
import java.util.Hashtable;
import java.util.Map;

/*
 * @Author: SourDumplings
 * @Date: 2019-09-06 10:46:09
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/word-pattern/
 */

class Solution
{
    public boolean wordPattern(String pattern, String str)
    {
        Map<Character, String> m1 = new Hashtable<>();
        Map<String, Character> m2 = new Hashtable<>();
        String[] sA = str.split(" ");
        int l = pattern.length();
        if (l != sA.length)
        {
            return false;
        }
        for (int i = 0; i < l; i++)
        {
            String s = sA[i];
            Character c = pattern.charAt(i);
            if (!m1.containsKey(c))
            {
                if (!m2.containsKey(s))
                {
                    m1.put(c, s);
                    m2.put(s, c);
                }
                else
                {
                    return false;
                }
            }
            else if (!m1.get(c).equals(s))
            {
                return false;
            }
        }
        return true;
    }
}

File: OJ practices/LeetCode题目集/290. Word Pattern(easy).cpp
/**
 * @file 290. Word Pattern(easy).cpp
 * @author SourDumplings (changzheng300@foxmail.com)
 * @brief 
 * @version 1.0.0
 * @date 2024-04-13
 * 
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 * https://leetcode.cn/problems/word-pattern/description/?envType=study-plan-v2&envId=top-interview-150
 * 
 */

class Solution
{
public:
    bool wordPattern(string pattern, string s)
    {
        unordered_map<char, string> m;
        unordered_map<string, char> rm;
        int l = pattern.length();
        int b = 0;
        for (int i = 0; i < l; i++)
        {
            if (s.length() <= b)
            {
                return false;
            }

            int e = b + 1;
            for (; e < s.length() && s[e] != ' '; ++e);
            string word = s.substr(b, e - b);
            b = e + 1;
            if (m.find(pattern[i]) == m.end())
            {
                if (rm.find(word) != rm.end())
                {
                    return false;
                }
                
                m[pattern[i]] = word;
                rm[word] = pattern[i];
            }
            else if (m.at(pattern[i]) != word)
            {
                return false;
            }
            
        }
        if (b < s.length() + 1)
        {
            return false;
        }
        return true;
    }
};

